utils/dataset_manifest/core.py

```diff
@@ -182,37 +193,69 @@ def step(self):
-                name, extension = os.path.splitext(img_name)
-                image_properties = {
-                    'name': name.replace('\\', '/'),
-                    'extension': extension,
-                }
+        name, extension = os.path.splitext(img_name)
+        image_properties = {
+            'name': name.replace('\\', '/'),
+            'extension': extension,
+        }
 
-                width, height = img.width, img.height
-                orientation = img.getexif().get(274, 1)
-                if orientation > 4:
-                    width, height = height, width
-                image_properties['width'] = width
-                image_properties['height'] = height
+        width, height = img.width, img.height
+        orientation = img.getexif().get(274, 1)
+        if orientation > 4:
+            width, height = height, width
+        image_properties['width'] = width
+        image_properties['height'] = height
 
-                if self._meta and img_name in self._meta:
-                    image_properties['meta'] = self._meta[img_name]
+        if self._meta and img_name in self._meta:
+            image_properties['meta'] = self._meta[img_name]
 
-                if self._use_image_hash:
-                    image_properties['checksum'] = md5_hash(img)
+        if self._use_image_hash:
+            image_properties['checksum'] = md5_hash(img)
 
-                yield image_properties
+        return image_properties
+
+    def __iter__(self):
+        if not self.are_sources_filtered:
+            if self.stop is not None:
+                index_range = range(self.start, self.stop, self.step)
+                idx = 0
+                for image in self._sources:
+                    if idx in index_range:
+                        yield self._get_img_properties(image)
+                    else:
+                        yield dict()
+                    idx += 1
+            elif self.start or self.step != 1:
+                idx = 0
+                for image in self._sources:
+                    if idx >= self.start and not (idx - self._start) % self.step:
+                        yield self._get_img_properties(image)
+                    else:
+                        yield dict()
             else:
-                yield dict()
+                for image in self._sources:
+                    yield self._get_img_properties(image)
+        else:
+            sources = iter(self._sources) if not self._is_generator_used else self._sources
+
+            assert self.start is not None
+            assert self.stop is not None
+            assert self.step is not None
+
+            index_range = range(self.start, self.stop, self.step)
+            for idx in range(self.stop):
+                if idx in index_range:
+                    image = next(sources)
+                    yield self._get_img_properties(image)
+                else:
+                    yield dict()
```

Comment:
Consider this as a function, which can be applied to an iterator and return an iterator.

Reply:
I do not see here a necessity to move this code into a separate function since now __iter__ does not look like a huge function

Reply:
Agree, now the function seems to be small enough.